vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py

```diff
@@ -0,0 +1,642 @@
+                else:
+                    raise RuntimeError("Transfer failed with state %s",
+                                       xfer_state)
+            if len(running_reqs) == 0:
+                done_req_ids.add(req_id)
+                del transfers[req_id]
+            else:
+                transfers[req_id] = running_reqs
+        return done_req_ids
+
+    def start_load_kv(self, metadata: NixlConnectorMetadata):
+        """
+        Start loading by triggering non-blocking nixl_xfer.
+        We check for these trnxs to complete in each step().
+        """
+        for req_id, meta in metadata.requests.items():
+            # NOTE: this is non-blocking
+            logger.debug("start_load_kv for request %s from remote engine %s. "
+                         "Num local_block_ids: %s. Num remote_block_ids: %s. ",
+                         req_id, meta.remote_engine_id,
+                         len(meta.local_block_ids),
+                         len(meta.remote_block_ids))
+            self._read_blocks(
+                local_block_ids=meta.local_block_ids,
+                remote_block_ids=meta.remote_block_ids,
+                dst_engine_id=meta.remote_engine_id,
+                request_id=req_id,
+            )
+
+    def _read_blocks(
+        self,
+        local_block_ids: list[int],
+        remote_block_ids: list[int],
+        dst_engine_id: str,
+        request_id: str,
+    ):
+        # NOTE(rob): having the staging blocks be on the READER side is
+        # not going to work well (since we will have to call rearrange tensors).
+        # after we detect the txn is complete (which means we cannot make the
+        # read trxn async easily). If we want to make "READ" happen cleanly,
+        # then we will need to have the staging blocks on the remote side.
+
+        # NOTE(rob): according to nvidia the staging blocks are used to
+        # saturate IB with heterogeneous TP sizes. We should remove the staging
+        # blocks until we are ready.
+
+        # NOTE(rob): we could potentially do the rearranging during the load_kv!
+
+        # Note(tms): The remote_block_ids only contain full computed blocks,
+        # while the local_block_ids are all blocks allocated for this request,
+        # so truncate the local_block_ids to account for this.
+        if len(remote_block_ids) < len(local_block_ids):
+            local_block_ids = local_block_ids[:len(remote_block_ids)]
+        assert len(local_block_ids) == len(remote_block_ids)
+        if len(local_block_ids) == 0:
+            return
+
+        # TODO: support TP multipliers.
+        tp_multiplier = 1
+        remote_block_descs_ids = self._get_block_descs_ids(
+            dst_engine_id, "all", remote_block_ids)
+        local_xfer_side_handle = self.src_xfer_side_handles[tp_multiplier]
+
+        # Read the data from the remote.
+        for i in range(tp_multiplier):
+            local_block_descs_ids = self._get_block_descs_ids(
```

Comment:
@tlrmchlsmth - I think this should be `self.engine_id` rather than `dst_engine_id`

Similar to: https://github.com/vllm-project/vllm/pull/16124/files#diff-876efa5533f5dcff3fba850e8684a47d53c112e287988957c115b11691374f4bR225